src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1144 @@
+from sage.misc.cachefunc import cached_method
+from sage.modules.free_module_element import vector
+from sage.rings.integer import Integer
+from sage.rings.integer_ring import ZZ
+from sage.rings.polynomial.polynomial_ring import polygen
+from sage.rings.polynomial.polynomial_ring_constructor import PolynomialRing
+from sage.rings.rational_field import QQ
+from sage.sets.non_negative_integers import NonNegativeIntegers
+from sage.sets.family import Family
+
+
+class IntegerValuedPolynomialRing(CombinatorialFreeModule):
+    r"""
+    The integer-valued polynomial ring over a base ring.
+
+    Integer-valued polynomial rings are commutative and associative
+    algebras, with a basis indexed by non-negative integers.
+
+    The basis used here is given by `B[i] = \binom{i+n}{i}` for `i \in \NN`.
+
+    The product of two monomials
+    `B[n_1] \cdot B[n_2]` is given by the sum over ...
+
+    There is a nice formula for the product (ref ?)
+
+    REFERENCES:
+
+    - :wikipedia:`Integer-valued polynomial`
+
+    INPUT:
+
+    - ``R`` -- ring
+
+    EXAMPLES::
+
+        sage: F = IntegerValuedPolynomialRing(QQ); F
+        Integer-Valued Polynomial Ring over Rational Field
+
+        sage: F.gen()
+        B[1]
+
+        sage: S = IntegerValuedPolynomialRing(ZZ); S
+        Integer-Valued Polynomial Ring over Integer Ring
+        sage: S.base_ring()
+        Integer Ring
+
+        sage: G = IntegerValuedPolynomialRing(S); G
+        Integer-Valued Polynomial Ring over Integer-Valued Polynomial
+        Ring over Integer Ring
+        sage: G.base_ring()
+        Integer-Valued Polynomial Ring over Integer Ring
+
+    Integer-valued polynomial rings commute with their base ring::
+
+        sage: K = IntegerValuedPolynomialRing(QQ)
+        sage: a = K.gen()
+        sage: K.is_commutative()
+        True
+        sage: L = IntegerValuedPolynomialRing(K)
+        sage: c = L.gen()
+        sage: L.is_commutative()
+        True
+        sage: s = a * c^3; s
+        B[1]*B[1] + (-6*B[1])*B[2] + 6*B[1]*B[3]
+        sage: parent(s)
+        Integer-Valued Polynomial Ring over Integer-Valued Polynomial
+        Ring over Rational Field
+
+    Integer-valued polynomial rings are commutative::
+
+        sage: c^3 * a == c * a * c * c
+        True
+
+    We can also manipulate elements in the basis and coerce elements from our
+    base field::
+
+        sage: F = IntegerValuedPolynomialRing(QQ)
+        sage: B = F.basis()
+        sage: B[2] * B[3]
+        3*B[3] - 12*B[4] + 10*B[5]
+        sage: 1 - B[2] * B[2] / 2
+        B[0] - 1/2*B[2] + 3*B[3] - 3*B[4]
+    """
+    @staticmethod
+    def __classcall_private__(cls, R):
+        """
+        Normalize input to ensure a unique representation.
+
+        EXAMPLES::
+
+            sage: F1 = IntegerValuedPolynomialRing(QQ)
+            sage: F2 = IntegerValuedPolynomialRing(QQ)
+            sage: F1 is F2
+            True
+        """
+        return super().__classcall__(cls, R)
```

Comment:
This can be removed since it is effectively a no-op. The `__init__` will filter out extra inputs.